opm/autodiff/BlackoilAmg.hpp

```diff
@@ -0,0 +1,727 @@
+#endif
+    const auto ISOLATED = Dune::Amg::AggregatesMap<typename G::VertexDescriptor>::ISOLATED;
+    using Vertex = typename G::VertexIterator;
+    Vertex vend = fineGraph.end();
+    for(Vertex vertex = fineGraph.begin(); vertex != vend; ++vertex) {
+        assert(aggregates[*vertex] != UNAGGREGATED);
+        put(visitedMap, *vertex, aggregates[*vertex]==ISOLATED);
+    }
+
+    Dune::Amg::SparsityBuilder<M> sparsityBuilder(coarseMatrix);
+
+    Dune::Amg::ConnectivityConstructor<G,Dune::Amg::SequentialInformation>
+        ::examine(fineGraph, visitedMap, pinfo, aggregates, sparsityBuilder);
+}
+
+} // end namespace Detail
+
+/**
+ * @brief A LevelTransferPolicy that uses aggregation to construct the coarse level system.
+ * @tparam Operator The type of the fine level operator.
+ * @tparam Criterion The criterion that describes the aggregation procedure.
+ * @tparam Communication The class that describes the communication pattern.
+ */
+template<class Operator, class Criterion, class Communication, std::size_t COMPONENT_INDEX>
+class OneComponentAggregationLevelTransferPolicy
+    : public Dune::Amg::LevelTransferPolicy<Operator, typename Detail::ScalarType<Operator>::value>
+{
+    typedef Dune::Amg::AggregatesMap<typename Operator::matrix_type::size_type> AggregatesMap;
+public:
+    using CoarseOperator = typename Detail::ScalarType<Operator>::value;
+    typedef Dune::Amg::LevelTransferPolicy<Operator,CoarseOperator> FatherType;
+    typedef Communication ParallelInformation;
+
+public:
+    OneComponentAggregationLevelTransferPolicy(const Criterion& crit, const Communication& comm)
+        : criterion_(crit), communication_(&const_cast<Communication&>(comm))
+    {}
+
+    void createCoarseLevelSystem(const Operator& fineOperator)
+    {
+        prolongDamp_ = 1;
+        typedef Dune::Amg::PropertiesGraphCreator<Operator> GraphCreator;
+        typedef typename GraphCreator::PropertiesGraph PropertiesGraph;
+        typedef typename GraphCreator::GraphTuple GraphTuple;
+
+        typedef typename PropertiesGraph::VertexDescriptor Vertex;
+
+        std::vector<bool> excluded(fineOperator.getmat().N(), false);
+
+        using OverlapFlags = Dune::NegateSet<typename ParallelInformation::OwnerSet>;
+        GraphTuple graphs = GraphCreator::create(fineOperator, excluded,
+                                                 *communication_, OverlapFlags());
+
+        aggregatesMap_.reset(new AggregatesMap(std::get<1>(graphs)->maxVertex()+1));
+
+        int noAggregates, isoAggregates, oneAggregates, skippedAggregates;
+        using std::get;
+        std::tie(noAggregates, isoAggregates, oneAggregates, skippedAggregates) =
+            aggregatesMap_->buildAggregates(fineOperator.getmat(), *(get<1>(graphs)),
+                                            criterion_, true);
+        std::cout<<"no aggregates="<<noAggregates<<" iso="<<isoAggregates<<" one="<<oneAggregates<<" skipped="<<skippedAggregates<<std::endl;
```

Comment:
cout logging wanted?

Reply:
comment out later and removed with next push